In [ ]:
import pandas as pd 

# Load the raw banking datasets
# ---------------------------------------------------------

customers = pd.read_csv("../data_raw/customers.csv")
branches = pd.read_csv("../data_raw/branches.csv")
loan_applications = pd.read_csv("../data_raw/loan_applications.csv")
loans = pd.read_csv("../data_raw/loans.csv")
payments = pd.read_csv("../data_raw/payments.csv")

Confirm that everything is loaded


In [ ]:
# Check the number of rows and columns in each table.


print("Customers:", customers.shape)
print("Branches:", branches.shape)
print("Loan Applications:", loan_applications.shape)
print("Loans:", loans.shape)
print("Payments:", payments.shape)

Customers: (12000, 11)
Branches: (25, 5)
Loan Applications: (20035, 8)
Loans: (14500, 11)
Payments: (194025, 9)


structure of each table


In [ ]:
# Display the column names for each table.


print("CUSTOMERS")
print(customers.columns.tolist())

print("\nBRANCHES")
print(branches.columns.tolist())

print("\nLOAN APPLICATIONS")
print(loan_applications.columns.tolist())

print("\nLOANS")
print(loans.columns.tolist())

print("\nPAYMENTS")
print(payments.columns.tolist())

CUSTOMERS
['Customer_ID', 'Customer_Name', 'Date_of_Birth', 'Gender', 'City', 'Region', 'Employment_Type', 'Monthly_Income', 'Customer_Segment', 'Customer_Since', 'Branch_ID']

BRANCHES
['Branch_ID', 'Branch_Name', 'City', 'Region', 'Branch_Open_Date']

LOAN APPLICATIONS
['Application_ID', 'Customer_ID', 'Branch_ID', 'Application_Date', 'Decision_Date', 'Loan_Type', 'Requested_Amount', 'Application_Status']

LOANS
['Loan_ID', 'Application_ID', 'Customer_ID', 'Branch_ID', 'Loan_Type', 'Approval_Date', 'Disbursement_Date', 'Loan_Amount', 'Interest_Rate', 'Tenure_Months', 'Loan_Status']

PAYMENTS
['Payment_ID', 'Loan_ID', 'Customer_ID', 'Due_Date', 'Payment_Date', 'Amount_Due', 'Amount_Paid', 'Payment_Status', 'Days_Late']


Quick Preview

In [ ]:
# View the first 5 records from each raw table.


print("CUSTOMERS")
display(customers.head())

print("BRANCHES")
display(branches.head())

print("LOAN APPLICATIONS")
display(loan_applications.head())

print("LOANS")
display(loans.head())

print("PAYMENTS")
display(payments.head())

CUSTOMERS


,Customer_ID,Customer_Name,Date_of_Birth,Gender,City,Region,Employment_Type,Monthly_Income,Customer_Segment,Customer_Since,Branch_ID
0,C000001,Priya Shah,1960-06-11,Female,Pune,West,Self-Employed,45300.0,Affluent,2024-09-11,B023
1,C000002,Sneha Jain,1967-07-03,Male,Kolkata,East,Salaried,40600.0,Affluent,2022-03-27,B018
2,C000003,Karan Das,1987-07-19,Female,Bhopal,Central,Salaried,109000.0,Premium,2018-06-15,B015
3,C000004,Aarav Rao,1983-10-14,Female,Ranchi,East,Business Owner,35600.0,Affluent,2021-10-02,B017
4,C000005,Aarav Sharma,1965-04-20,Female,Bhubaneswar,East,Salaried,43900.0,Affluent,2020-12-06,B007


BRANCHES


,Branch_ID,Branch_Name,City,Region,Branch_Open_Date
0,B001,Delhi 1 Branch,Delhi,North,2018-08-12
1,B002,Chennai 2 Branch,Chennai,South,2023-04-19
2,B003,Patna 3 Branch,Patna,East,2022-06-25
3,B004,Surat 4 Branch,Surat,West,2021-01-02
4,B005,Kanpur 5 Branch,Kanpur,Central,2020-12-18


LOAN APPLICATIONS


,Application_ID,Customer_ID,Branch_ID,Application_Date,Decision_Date,Loan_Type,Requested_Amount,Application_Status
0,A0000001,C002406,B013,2024-01-06 03:46:25,2024-01-12 03:46:25,Business Loan,2675000.0,Rejected
1,A0000002,C001218,B017,2024-01-26 05:05:12,2024-02-08 05:05:12,Personal Loan,695000.0,Approved
2,A0000003,C002386,B007,2023-04-13 10:02:20,2023-04-23 10:02:20,Business Loan,2175000.0,Approved
3,A0000004,C006435,B017,2025-05-08 05:52:35,2025-05-12 05:52:35,Personal Loan,965000.0,Approved
4,A0000005,C001389,B019,2023-11-01 18:06:31,2023-11-11 18:06:31,Auto Loan,1535000.0,Approved


LOANS


,Loan_ID,Application_ID,Customer_ID,Branch_ID,Loan_Type,Approval_Date,Disbursement_Date,Loan_Amount,Interest_Rate,Tenure_Months,Loan_Status
0,L0000001,A0010472,C000925,B005,Auto Loan,2024-02-05 14:33:22,2024-03-01 14:33:22,630000.0,12.60,48.0,Closed
1,L0000002,A0000218,C001124,B001,Business Loan,2023-04-03 12:54:12,2023-04-07 12:54:12,2395000.0,12.43,48.0,Closed
2,L0000003,A0018819,C008034,B020,Education Loan,2024-07-03 06:42:29,2024-07-27 06:42:29,585000.0,8.58,36.0,Active
3,L0000004,A0011215,C007166,B002,Business Loan,2023-03-01 15:30:15,2023-03-12 15:30:15,630000.0,12.01,24.0,Active
4,L0000005,A0005187,C002683,B001,Home Loan,2024-07-24 11:19:02,2024-08-02 11:19:02,4240000.0,9.56,300.0,Active


PAYMENTS


,Payment_ID,Loan_ID,Customer_ID,Due_Date,Payment_Date,Amount_Due,Amount_Paid,Payment_Status,Days_Late
0,P00000001,L0000001,C000925,2024-04-01 14:33:22,2024-04-03 14:33:22,17500.0,17712.27,Paid,2.0
1,P00000002,L0000001,C000925,2024-05-01 14:33:22,2024-06-01 14:33:22,17500.0,16814.07,Late,31.0
2,P00000003,L0000001,C000925,2024-06-01 14:33:22,2024-07-03 14:33:22,17500.0,16429.15,Late,32.0
3,P00000004,L0000001,C000925,2024-07-01 14:33:22,2024-07-02 14:33:22,17500.0,17659.65,Paid,1.0
4,P00000005,L0000001,C000925,2024-08-01 14:33:22,2024-08-02 14:33:22,17500.0,17835.15,Paid,1.0


Basic Technical info

In [ ]:
# ---------------------------------------------------------
# Inspect data types and non-null counts

print("========== CUSTOMERS ==========")
customers.info()

print("\n========== BRANCHES ==========")
branches.info()

print("\n========== LOAN APPLICATIONS ==========")
loan_applications.info()

print("\n========== LOANS ==========")
loans.info()

print("\n========== PAYMENTS ==========")
payments.info()

========== CUSTOMERS ==========
<class 'pandas.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Customer_ID       12000 non-null  str    
 1   Customer_Name     12000 non-null  str    
 2   Date_of_Birth     12000 non-null  str    
 3   Gender            12000 non-null  str    
 4   City              11964 non-null  str    
 5   Region            12000 non-null  str    
 6   Employment_Type   11928 non-null  str    
 7   Monthly_Income    11952 non-null  float64
 8   Customer_Segment  12000 non-null  str    
 9   Customer_Since    12000 non-null  str    
 10  Branch_ID         12000 non-null  str    
dtypes: float64(1), str(10)
memory usage: 1.0 MB

========== BRANCHES ==========
<class 'pandas.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  ---

missing values in each table

In [ ]:
# ---------------------------------------------------------
# Check missing values in each table


print("CUSTOMERS - Missing Values")
print(customers.isna().sum())

print("\nBRANCHES - Missing Values")
print(branches.isna().sum())

print("\nLOAN APPLICATIONS - Missing Values")
print(loan_applications.isna().sum())

print("\nLOANS - Missing Values")
print(loans.isna().sum())

print("\nPAYMENTS - Missing Values")
print(payments.isna().sum())

CUSTOMERS - Missing Values
Customer_ID          0
Customer_Name        0
Date_of_Birth        0
Gender               0
City                36
Region               0
Employment_Type     72
Monthly_Income      48
Customer_Segment     0
Customer_Since       0
Branch_ID            0
dtype: int64

BRANCHES - Missing Values
Branch_ID           0
Branch_Name         0
City                0
Region              0
Branch_Open_Date    0
dtype: int64

LOAN APPLICATIONS - Missing Values
Application_ID        0
Customer_ID           0
Branch_ID             0
Application_Date      0
Decision_Date         0
Loan_Type             0
Requested_Amount      0
Application_Status    0
dtype: int64

LOANS - Missing Values
Loan_ID               0
Application_ID        0
Customer_ID           0
Branch_ID             0
Loan_Type             0
Approval_Date         0
Disbursement_Date     0
Loan_Amount           0
Interest_Rate        25
Tenure_Months        25
Loan_Status           0
dtype: int64

PAYMENTS - Mis

Customer table cleaning

In [ ]:
customers_clean = customers.copy()
# Standardize Employment_Type capitalization.
# "Salaried" and "salaried" should represent the same category.
customers_clean["Employment_Type"] = (
    customers_clean["Employment_Type"]
    .str.strip()
    .str.title()
)

# Convert customer_since from text/object into a proper datetime column.
# errors='coerce' converts invalid dates into NaT instead of crashing.
customers_clean["Customer_Since"] = pd.to_datetime(
    customers_clean["Customer_Since"],
    errors="coerce"
)

# City is a categorical field.
# so we explicitly label missing information as "Unknown".
customers_clean["City"] = customers_clean["City"].fillna("Unknown")

# Employment type is also categorical.
# "Unknown" preserves the fact that the source did not provide this information.
customers_clean["Employment_Type"] = (
    customers_clean["Employment_Type"]
    .fillna("Unknown")
)


# Check the result.
customers_clean[[
    "Customer_Since",
    "City",
    "Employment_Type"
]].isna().sum()

Customer_Since     0
City               0
Employment_Type    0
dtype: int64

monthly income inspection

In [29]:
customers_clean["Employment_Type"].value_counts(dropna=False)

Employment_Type
Salaried          5713
Self-Employed     2568
Business Owner    1694
Retired           1230
Student            723
Unknown             72
Name: count, dtype: int64

In [ ]:
# Fill missing income using the median income within each employment type.

customers_clean["Monthly_Income"] = (
    customers_clean.groupby("Employment_Type")["Monthly_Income"]
    .transform(lambda x: x.fillna(x.median()))
)

In [31]:
customers_clean["Monthly_Income"].isna().sum()

np.int64(0)

Payments table

In [32]:
payments[payments["Days_Late"].isna()].head()

,Payment_ID,Loan_ID,Customer_ID,Due_Date,Payment_Date,Amount_Due,Amount_Paid,Payment_Status,Days_Late
32,P00000033,L0000002,C001124,2024-10-07 12:54:12,NaN,66527.78,0.0,Missed,NaN
85,P00000086,L0000005,C002683,2025-06-02 11:19:02,NaN,117777.78,0.0,Missed,NaN
105,P00000106,L0000007,C000463,2024-10-20 22:42:17,NaN,31875.00,0.0,Missed,NaN
172,P00000173,L0000013,C003229,2024-11-26 19:12:30,NaN,45972.22,0.0,Missed,NaN
192,P00000193,L0000014,C005921,2024-06-03 11:09:41,NaN,20555.56,0.0,Missed,NaN


Payments — Date Cleaning & Business-Rule Validation

In [ ]:


# Convert both date columns to proper datetime format.
# errors='coerce' converts genuinely invalid date values to NaT

payments["Due_Date"] = pd.to_datetime(
    payments["Due_Date"],
    errors="coerce"
)

payments["Payment_Date"] = pd.to_datetime(
    payments["Payment_Date"],
    errors="coerce"
)

# Check whether any invalid/missing dates were created during conversion.

print("Missing Due_Date:", payments["Due_Date"].isna().sum())
print("Missing Payment_Date:", payments["Payment_Date"].isna().sum())

Missing Due_Date: 0
Missing Payment_Date: 9561


Payment_Date is missing for 9,561 records, but these records represent
missed payments. Since no payment occurred, a Payment_Date is not applicable.
Therefore, these NaN values are retained rather than artificially filled.

Business-rule validation

In [34]:

# Rule 1:
# A Paid payment should have a Payment_Date.
paid_without_date = payments[
    (payments["Payment_Status"] == "Paid") &
    (payments["Payment_Date"].isna())
]

# Rule 2:
# A Paid payment should not have a Payment_Date earlier than Due_Date.
paid_before_due = payments[
    (payments["Payment_Status"] == "Paid") &
    (payments["Payment_Date"] < payments["Due_Date"])
]

# Rule 3:
# A Missed payment should not have an actual amount paid.
missed_with_payment = payments[
    (payments["Payment_Status"] == "Missed") &
    (payments["Amount_Paid"] > 0)
]

# Rule 4:
# A Missed payment should not have a Payment_Date.
missed_with_date = payments[
    (payments["Payment_Status"] == "Missed") &
    (payments["Payment_Date"].notna())
]

print("Paid payments without Payment_Date:", len(paid_without_date))
print("Paid payments before Due_Date:", len(paid_before_due))
print("Missed payments with Amount_Paid > 0:", len(missed_with_payment))
print("Missed payments with Payment_Date:", len(missed_with_date))

Paid payments without Payment_Date: 0
Paid payments before Due_Date: 0
Missed payments with Amount_Paid > 0: 0
Missed payments with Payment_Date: 0


In [35]:
# Keep the raw Payments table unchanged.
# All cleaning operations will be performed on this copy.

payments_clean = payments.copy()

# Convert date fields to proper datetime format.
payments_clean["Due_Date"] = pd.to_datetime(
    payments_clean["Due_Date"],
    errors="coerce"
)

payments_clean["Payment_Date"] = pd.to_datetime(
    payments_clean["Payment_Date"],
    errors="coerce"
)

cross check

In [36]:
loan_applications.info()

<class 'pandas.DataFrame'>
RangeIndex: 20035 entries, 0 to 20034
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Application_ID      20035 non-null  str    
 1   Customer_ID         20035 non-null  str    
 2   Branch_ID           20035 non-null  str    
 3   Application_Date    20035 non-null  str    
 4   Decision_Date       20035 non-null  str    
 5   Loan_Type           20035 non-null  str    
 6   Requested_Amount    20035 non-null  float64
 7   Application_Status  20035 non-null  str    
dtypes: float64(1), str(7)
memory usage: 1.2 MB


 Loan Applications — Date Cleaning & Validation


In [37]:

# Convert application and decision dates to datetime.
# These columns are currently stored as strings, so conversion
# is required for reliable date calculations and comparisons.
loan_applications["Application_Date"] = pd.to_datetime(
    loan_applications["Application_Date"],
    errors="coerce"
)

loan_applications["Decision_Date"] = pd.to_datetime(
    loan_applications["Decision_Date"],
    errors="coerce"
)

# Check whether any invalid dates became NaT during conversion.
print(
    "Invalid Application_Date:",
    loan_applications["Application_Date"].isna().sum()
)

print(
    "Invalid Decision_Date:",
    loan_applications["Decision_Date"].isna().sum()
)

Invalid Application_Date: 0
Invalid Decision_Date: 0


In [38]:
# A decision cannot logically occur before the application date.
invalid_decision_dates = loan_applications[
    loan_applications["Decision_Date"] <
    loan_applications["Application_Date"]
]

print(
    "Applications with Decision_Date before Application_Date:",
    len(invalid_decision_dates)
)

Applications with Decision_Date before Application_Date: 9


In [39]:
invalid_decision_dates[
    [
        "Application_ID",
        "Customer_ID",
        "Application_Date",
        "Decision_Date",
        "Loan_Type",
        "Requested_Amount",
        "Application_Status"
    ]
]

,Application_ID,Customer_ID,Application_Date,Decision_Date,Loan_Type,Requested_Amount,Application_Status
3201,A0003202,C003183,2023-05-01,2023-01-19 18:24:20,Personal Loan,425000.0,Rejected
3459,A0003460,C004339,2024-10-08,2024-08-24 03:35:18,Personal Loan,485000.0,Approved
5266,A0005267,C005597,2023-07-01,2023-01-16 05:29:56,Auto Loan,1110000.0,Rejected
5425,A0005426,C011472,2024-11-09,2024-09-20 08:21:35,Auto Loan,1330000.0,Approved
8469,A0008470,C010467,2023-11-09,2023-09-13 09:18:52,Business Loan,515000.0,Rejected
10016,A0010017,C009053,2025-08-05,2025-05-18 13:49:26,Home Loan,2345000.0,Approved
13137,A0013138,C009769,2023-03-02,2023-02-08 15:00:01,Auto Loan,1795000.0,Approved
15229,A0015230,C011815,2024-06-03,2024-03-20 04:28:15,Business Loan,3085000.0,Approved
16871,A0016872,C008908,2025-06-01,2025-01-12 15:51:31,Personal Loan,130000.0,Rejected


In [40]:
# Flag applications where the decision date occurs before
# the application date.
loan_applications["Invalid_Date_Flag"] = (
    loan_applications["Decision_Date"] <
    loan_applications["Application_Date"]
)

In [41]:


loan_applications_clean = loan_applications[
    ~loan_applications["Invalid_Date_Flag"]
].copy()

In [ ]:
# Application_ID should uniquely identify each application.


loan_applications_clean["Application_ID"].duplicated().sum()

np.int64(35)

In [43]:
# Remove duplicate application records.
# Keep the first occurrence and discard repeated records.
loan_applications_clean = loan_applications_clean.drop_duplicates(
    subset="Application_ID",
    keep="first"
)

In [44]:
loan_applications_clean["Application_ID"].duplicated().sum()

np.int64(0)

In [45]:
loan_applications_clean.isna().sum()

Application_ID        0
Customer_ID           0
Branch_ID             0
Application_Date      0
Decision_Date         0
Loan_Type             0
Requested_Amount      0
Application_Status    0
Invalid_Date_Flag     0
dtype: int64

In [46]:
# Standardize Application_Status so the same business category
# is not represented with different capitalization.
loan_applications_clean["Application_Status"] = (
    loan_applications_clean["Application_Status"]
    .str.strip()
    .str.title()
)

loan_applications_clean["Application_Status"].value_counts()

Application_Status
Approved    15821
Rejected     4170
Name: count, dtype: int64

In [47]:
loans.info()

<class 'pandas.DataFrame'>
RangeIndex: 14500 entries, 0 to 14499
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            14500 non-null  str    
 1   Application_ID     14500 non-null  str    
 2   Customer_ID        14500 non-null  str    
 3   Branch_ID          14500 non-null  str    
 4   Loan_Type          14500 non-null  str    
 5   Approval_Date      14500 non-null  str    
 6   Disbursement_Date  14500 non-null  str    
 7   Loan_Amount        14500 non-null  float64
 8   Interest_Rate      14475 non-null  float64
 9   Tenure_Months      14475 non-null  float64
 10  Loan_Status        14500 non-null  str    
dtypes: float64(3), str(8)
memory usage: 1.2 MB


In [ ]:
# Create a separate clean copy so the original raw loans data stays unchanged.
loans_clean = loans.copy()

# Convert loan dates from text to datetime for reliable date calculations 
loans_clean["Approval_Date"] = pd.to_datetime(
    loans_clean["Approval_Date"],
    errors="coerce"
)

loans_clean["Disbursement_Date"] = pd.to_datetime(
    loans_clean["Disbursement_Date"],
    errors="coerce"
)

In [49]:
print("Invalid Approval_Date:", loans_clean["Approval_Date"].isna().sum())
print("Invalid Disbursement_Date:", loans_clean["Disbursement_Date"].isna().sum())

Invalid Approval_Date: 0
Invalid Disbursement_Date: 0


In [50]:
# A loan cannot logically be disbursed before it is approved.
invalid_loan_dates = loans_clean[
    loans_clean["Disbursement_Date"] < loans_clean["Approval_Date"]
]

print("Loans with Disbursement_Date before Approval_Date:",
      len(invalid_loan_dates))

Loans with Disbursement_Date before Approval_Date: 12


In [ ]:
# Flag loans where disbursement happened before approval.

loans_clean["Invalid_Date_Flag"] = (
    loans_clean["Disbursement_Date"] <
    loans_clean["Approval_Date"]
)

print(loans_clean["Invalid_Date_Flag"].sum())

12


In [52]:
# Exclude records with impossible loan dates.
# We retain the raw loans table separately for traceability.
loans_clean = loans_clean[
    ~loans_clean["Invalid_Date_Flag"]
].copy()

In [53]:
# Confirm that no invalid date sequences remain.
print(
    "Invalid loan date records remaining:",
    (
        loans_clean["Disbursement_Date"] <
        loans_clean["Approval_Date"]
    ).sum()
)

Invalid loan date records remaining: 0


In [ ]:
# Show the loan types for records where Interest_Rate or Tenure_Months is missing.


loans_clean[
    loans_clean["Interest_Rate"].isna() |
    loans_clean["Tenure_Months"].isna()
][["Loan_ID", "Loan_Type", "Interest_Rate", "Tenure_Months"]]

,Loan_ID,Loan_Type,Interest_Rate,Tenure_Months
1441,L0001442,Auto Loan,11.75,NaN
1568,L0001569,Auto Loan,11.66,NaN
1617,L0001618,Auto Loan,NaN,24.0
1756,L0001757,Auto Loan,12.74,NaN
1975,L0001976,Personal Loan,NaN,24.0
1983,L0001984,Personal Loan,NaN,48.0
2213,L0002214,Business Loan,NaN,84.0
2411,L0002412,Business Loan,12.77,NaN
2453,L0002454,Home Loan,7.07,NaN
2864,L0002865,Home Loan,NaN,120.0


In [ ]:
# Fill missing loan attributes using the median for each Loan_Type.
# Using Loan_Type makes the imputation business-aware:

loans_clean["Interest_Rate"] = (
    loans_clean.groupby("Loan_Type")["Interest_Rate"]
    .transform(lambda x: x.fillna(x.median()))
)

loans_clean["Tenure_Months"] = (
    loans_clean.groupby("Loan_Type")["Tenure_Months"]
    .transform(lambda x: x.fillna(x.median()))
)

In [56]:
# Confirm that no missing values remain in these two loan attributes.
print(
    "Missing Interest_Rate:",
    loans_clean["Interest_Rate"].isna().sum()
)

print(
    "Missing Tenure_Months:",
    loans_clean["Tenure_Months"].isna().sum()
)

Missing Interest_Rate: 0
Missing Tenure_Months: 0


In [57]:
# A disbursed loan should have a positive loan amount.
# Zero or negative amounts would violate a basic banking business rule.

invalid_loan_amount = loans_clean[
    loans_clean["Loan_Amount"] <= 0
]

print("Loans with non-positive Loan_Amount:", len(invalid_loan_amount))

Loans with non-positive Loan_Amount: 4


In [ ]:
# Flag loans with a zero or negative Loan_Amount.


loans_clean["Invalid_Amount_Flag"] = (
    loans_clean["Loan_Amount"] <= 0
)

print("Invalid Loan_Amount records:",
      loans_clean["Invalid_Amount_Flag"].sum())

Invalid Loan_Amount records: 4


In [ ]:
# Exclude invalid financial records from the clean analytical table.

loans_clean = loans_clean[
    ~loans_clean["Invalid_Amount_Flag"]
].copy()

In [60]:
print(
    "Non-positive Loan_Amount remaining:",
    (loans_clean["Loan_Amount"] <= 0).sum()
)

Non-positive Loan_Amount remaining: 0


In [ ]:
# ---------------------------------------------------------
# Loans — Final categorical & ID validation
# ---------------------------------------------------------

# Standardize text fields:
# strip() removes accidental leading/trailing spaces,
# while title() makes category capitalization consistent.
for col in ["Loan_Type", "Loan_Status"]:
    loans_clean[col] = (
        loans_clean[col]
        .str.strip()
        .str.title()
    )


# Check whether Loan_Status contains unexpected categories.
# We expect statuses such as Active, Closed, Defaulted, etc.,

print("Loan Status values:")
print(loans_clean["Loan_Status"].value_counts())


# Loan_ID should uniquely identify each loan.
print("\nDuplicate Loan_IDs:",
      loans_clean["Loan_ID"].duplicated().sum())


# Application_ID should also be unique within the loan table
# because one approved application should create one loan.
print("Duplicate Application_IDs:",
      loans_clean["Application_ID"].duplicated().sum())

Loan Status values:
Loan_Status
Active     8857
Closed     4590
Default    1037
Name: count, dtype: int64

Duplicate Loan_IDs: 0
Duplicate Application_IDs: 0


In [62]:
branches.info()

<class 'pandas.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   Branch_ID         25 non-null     str  
 1   Branch_Name       25 non-null     str  
 2   City              25 non-null     str  
 3   Region            25 non-null     str  
 4   Branch_Open_Date  25 non-null     str  
dtypes: str(5)
memory usage: 1.1 KB


In [63]:
# ---------------------------------------------------------
# Branches — Date & Key Validation
# ---------------------------------------------------------

# Create a clean copy so the raw Branches table remains unchanged.
branches_clean = branches.copy()

# Convert Branch_Open_Date from text to datetime.
branches_clean["Branch_Open_Date"] = pd.to_datetime(
    branches_clean["Branch_Open_Date"],
    errors="coerce"
)

# Check whether any invalid dates were introduced.
print(
    "Invalid Branch_Open_Date:",
    branches_clean["Branch_Open_Date"].isna().sum()
)

# Branch_ID should uniquely identify each branch.
print(
    "Duplicate Branch_IDs:",
    branches_clean["Branch_ID"].duplicated().sum()
)

Invalid Branch_Open_Date: 0
Duplicate Branch_IDs: 0


integrity check

In [64]:
# ---------------------------------------------------------
# Cross-table ID validation
# ---------------------------------------------------------
# These checks verify that relationships between our banking
# tables are valid before we begin business analysis.
# ---------------------------------------------------------

# Customers referenced by Loan Applications should exist
missing_customer_in_applications = (
    ~loan_applications_clean["Customer_ID"].isin(customers_clean["Customer_ID"])
).sum()

# Applications referenced by Loans should exist
missing_application_in_loans = (
    ~loans_clean["Application_ID"].isin(
        loan_applications_clean["Application_ID"]
    )
).sum()

# Customers referenced by Loans should exist
missing_customer_in_loans = (
    ~loans_clean["Customer_ID"].isin(customers_clean["Customer_ID"])
).sum()

# Loans referenced by Payments should exist
missing_loan_in_payments = (
    ~payments_clean["Loan_ID"].isin(loans_clean["Loan_ID"])
).sum()

# Branches referenced by Applications should exist
missing_branch_in_applications = (
    ~loan_applications_clean["Branch_ID"].isin(branches_clean["Branch_ID"])
).sum()

# Branches referenced by Loans should exist
missing_branch_in_loans = (
    ~loans_clean["Branch_ID"].isin(branches_clean["Branch_ID"])
).sum()


print("Customer IDs missing from Applications:", missing_customer_in_applications)
print("Application IDs missing from Loans:", missing_application_in_loans)
print("Customer IDs missing from Loans:", missing_customer_in_loans)
print("Loan IDs missing from Payments:", missing_loan_in_payments)
print("Branch IDs missing from Applications:", missing_branch_in_applications)
print("Branch IDs missing from Loans:", missing_branch_in_loans)

Customer IDs missing from Applications: 0
Application IDs missing from Loans: 4
Customer IDs missing from Loans: 0
Loan IDs missing from Payments: 237
Branch IDs missing from Applications: 0
Branch IDs missing from Loans: 0


In [ ]:
# Identify loans whose Application_ID does not exist in the cleaned Loan Applications table.

invalid_application_links = loans_clean[
    ~loans_clean["Application_ID"].isin(
        loan_applications_clean["Application_ID"]
    )
]

invalid_application_links[
    [
        "Loan_ID",
        "Application_ID",
        "Customer_ID",
        "Loan_Type",
        "Loan_Amount",
        "Loan_Status"
    ]
]

,Loan_ID,Application_ID,Customer_ID,Loan_Type,Loan_Amount,Loan_Status
7907,L0007908,A0010017,C009053,Home Loan,2110000.0,Closed
11696,L0011697,A0013138,C009769,Auto Loan,1465000.0,Closed
13534,L0013535,A0015230,C011815,Business Loan,3085000.0,Active
13803,L0013804,A0003460,C004339,Personal Loan,435000.0,Closed


In [ ]:
# Remove loans whose Application_ID no longer exists
# These 4 records became orphaned because their corresponding
# applications were excluded due to invalid date chronology.

loans_clean = loans_clean[
    loans_clean["Application_ID"].isin(
        loan_applications_clean["Application_ID"]
    )
].copy()

In [67]:
print(
    "Application IDs missing from Loans:",
    (
        ~loans_clean["Application_ID"].isin(
            loan_applications_clean["Application_ID"]
        )
    ).sum()
)

Application IDs missing from Loans: 0


In [ ]:
# Keep only payments whose Loan_ID exists in the cleaned Loans table.

payments_clean = payments_clean[
    payments_clean["Loan_ID"].isin(loans_clean["Loan_ID"])
].copy()

In [69]:
print(
    "Loan IDs missing from Payments after cleaning:",
    (
        ~payments_clean["Loan_ID"].isin(loans_clean["Loan_ID"])
    ).sum()
)

Loan IDs missing from Payments after cleaning: 0


In [70]:
import os

# Create the data_clean folder if it does not already exist.
os.makedirs("../data_clean", exist_ok=True)

# Save each cleaned table as a separate CSV.
customers_clean.to_csv("../data_clean/customers_clean.csv", index=False)
branches_clean.to_csv("../data_clean/branches_clean.csv", index=False)
loan_applications_clean.to_csv("../data_clean/loan_applications_clean.csv", index=False)
loans_clean.to_csv("../data_clean/loans_clean.csv", index=False)
payments_clean.to_csv("../data_clean/payments_clean.csv", index=False)

print("All cleaned datasets saved successfully.")

All cleaned datasets saved successfully.


In [71]:
print("Customers:", len(customers), "→", len(customers_clean))
print("Branches:", len(branches), "→", len(branches_clean))
print("Applications:", len(loan_applications), "→", len(loan_applications_clean))
print("Loans:", len(loans), "→", len(loans_clean))
print("Payments:", len(payments), "→", len(payments_clean))

Customers: 12000 → 12000
Branches: 25 → 25
Applications: 20035 → 19991
Loans: 14500 → 14480
Payments: 194025 → 193737


In [72]:
import sqlite3

# Create an in-memory SQL database.
# This allows us to practice SQL directly on our Pandas DataFrames
# without needing to load the data into MySQL first.
conn = sqlite3.connect(":memory:")

# Register the cleaned Loans table as a SQL table.
loans_clean.to_sql(
    "loans",
    conn,
    index=False,
    if_exists="replace"
)

# Business Question:
# Which loan types have the highest loan volume and total loan amount?

query = """
SELECT
    Loan_Type,
    COUNT(Loan_ID) AS Loan_Count,
    ROUND(SUM(Loan_Amount), 2) AS Total_Loan_Amount,
    AVG(Loan_Amount) AS Avg_Loan_Amount
FROM loans
GROUP BY Loan_Type
ORDER BY Total_Loan_Amount DESC;
"""

portfolio = pd.read_sql(query, conn)

portfolio

,Loan_Type,Loan_Count,Total_Loan_Amount,Avg_Loan_Amount
0,Home Loan,2932,8.649205e+09,2.949933e+06
1,Business Loan,2508,4.193080e+09,1.671882e+06
2,Auto Loan,2709,2.360170e+09,8.712329e+05
3,Personal Loan,4919,2.254585e+09,4.583421e+05
4,Education Loan,1412,7.903450e+08,5.597344e+05


How much money is currently Active, Closed, or in Default, and how many loans are in each status?

In [73]:
query = """
SELECT
    Loan_Status,
    COUNT(Loan_ID) AS Loan_Count,
    ROUND(SUM(Loan_Amount), 2) AS Total_Loan_Amount,
    ROUND(AVG(Loan_Amount), 2) AS Avg_Loan_Amount
FROM loans
GROUP BY Loan_Status
ORDER BY Total_Loan_Amount DESC;
"""

loan_status = pd.read_sql(query, conn)

loan_status

,Loan_Status,Loan_Count,Total_Loan_Amount,Avg_Loan_Amount
0,Active,8856,1.107990e+10,1251117.89
1,Closed,4587,5.865980e+09,1278827.12
2,Default,1037,1.301505e+09,1255067.50


Which loan types have the highest default exposure?

In [74]:
query = """
SELECT
    Loan_Type,
    COUNT(Loan_ID) AS Default_Loan_Count,
    ROUND(SUM(Loan_Amount), 2) AS Default_Amount
FROM loans
WHERE Loan_Status = 'Default'
GROUP BY Loan_Type
ORDER BY Default_Amount DESC;
"""

default_by_type = pd.read_sql(query, conn)

default_by_type

,Loan_Type,Default_Loan_Count,Default_Amount
0,Home Loan,214,630605000.0
1,Business Loan,174,286885000.0
2,Auto Loan,203,174825000.0
3,Personal Loan,359,163555000.0
4,Education Loan,87,45635000.0


In [75]:
query = """
SELECT
    Loan_Type,

    -- Total number of loans for each loan type
    COUNT(Loan_ID) AS Total_Loans,

    -- Number of loans that are in Default
    SUM(CASE
        WHEN Loan_Status = 'Default' THEN 1
        ELSE 0
    END) AS Default_Loans,

    -- Percentage of loans that are in Default
    ROUND(
        100.0 * SUM(CASE
            WHEN Loan_Status = 'Default' THEN 1
            ELSE 0
        END) / COUNT(Loan_ID),
        2
    ) AS Default_Rate,

    -- Total money currently exposed through defaulted loans
    ROUND(
        SUM(CASE
            WHEN Loan_Status = 'Default' THEN Loan_Amount
            ELSE 0
        END),
        2
    ) AS Default_Amount

FROM loans

GROUP BY Loan_Type

ORDER BY Default_Rate DESC;
"""

default_analysis = pd.read_sql(query, conn)

default_analysis

,Loan_Type,Total_Loans,Default_Loans,Default_Rate,Default_Amount
0,Auto Loan,2709,203,7.49,174825000.0
1,Personal Loan,4919,359,7.30,163555000.0
2,Home Loan,2932,214,7.30,630605000.0
3,Business Loan,2508,174,6.94,286885000.0
4,Education Loan,1412,87,6.16,45635000.0


Which customer employment groups have the largest loan exposure and default risk?

In [76]:
# Register the cleaned Customers table in the same SQLite connection.
# The SQL JOIN below can then connect customers to loans using Customer_ID.

customers_clean.to_sql(
    "customers",
    conn,
    index=False,
    if_exists="replace"
)

# Now run the customer-segment analysis.
query = """
SELECT
    c.Employment_Type,

    -- Number of loans for each employment group
    COUNT(l.Loan_ID) AS Total_Loans,

    -- Total loan exposure for each employment group
    ROUND(SUM(l.Loan_Amount), 2) AS Total_Loan_Amount,

    -- Number of defaulted loans
    SUM(
        CASE
            WHEN l.Loan_Status = 'Default' THEN 1
            ELSE 0
        END
    ) AS Default_Loans,

    -- Default percentage within each employment group
    ROUND(
        100.0 * SUM(
            CASE
                WHEN l.Loan_Status = 'Default' THEN 1
                ELSE 0
            END
        ) / COUNT(l.Loan_ID),
        2
    ) AS Default_Rate

FROM loans l

JOIN customers c
    ON l.Customer_ID = c.Customer_ID

GROUP BY c.Employment_Type

ORDER BY Total_Loan_Amount DESC;
"""

employment_analysis = pd.read_sql(query, conn)

employment_analysis

,Employment_Type,Total_Loans,Total_Loan_Amount,Default_Loans,Default_Rate
0,Salaried,6872,8.681585e+09,506,7.36
1,Self-Employed,3135,3.937790e+09,205,6.54
2,Business Owner,2021,2.543485e+09,155,7.67
3,Retired,1504,1.887855e+09,102,6.78
4,Student,855,1.085810e+09,62,7.25
5,Unknown,93,1.108600e+08,7,7.53


In [77]:
# Register the cleaned Branches table so SQL can JOIN it with Loans.
branches_clean.to_sql(
    "branches",
    conn,
    index=False,
    if_exists="replace"
)

query = """
SELECT
    b.Branch_ID,
    b.Branch_Name,
    b.City,
    b.Region,

    -- Number of loans handled by the branch
    COUNT(l.Loan_ID) AS Loan_Count,

    -- Total amount of loans handled by the branch
    ROUND(SUM(l.Loan_Amount), 2) AS Total_Loan_Amount,

    -- Average loan size at the branch
    ROUND(AVG(l.Loan_Amount), 2) AS Avg_Loan_Amount

FROM loans l

JOIN branches b
    ON l.Branch_ID = b.Branch_ID

GROUP BY
    b.Branch_ID,
    b.Branch_Name,
    b.City,
    b.Region

ORDER BY Total_Loan_Amount DESC;
"""

branch_analysis = pd.read_sql(query, conn)

branch_analysis

,Branch_ID,Branch_Name,City,Region,Loan_Count,Total_Loan_Amount,Avg_Loan_Amount
0,B025,Kanpur 25 Branch,Kanpur,Central,619,791370000.0,1278465.27
1,B008,Patna 8 Branch,Patna,East,606,775285000.0,1279348.18
2,B014,Surat 14 Branch,Surat,West,588,771085000.0,1311369.05
3,B022,Chennai 22 Branch,Chennai,South,618,770305000.0,1246448.22
4,B002,Chennai 2 Branch,Chennai,South,583,764400000.0,1311149.23
5,B021,Delhi 21 Branch,Delhi,North,582,755500000.0,1298109.97
6,B006,Delhi 6 Branch,Delhi,North,569,746920000.0,1312688.93
7,B019,Surat 19 Branch,Surat,West,577,746085000.0,1293041.59
8,B020,Kanpur 20 Branch,Kanpur,Central,594,739570000.0,1245067.34
9,B023,Patna 23 Branch,Patna,East,554,739475000.0,1334792.42


How frequently are payments late, and which loan types have the highest late-payment rate?

In [78]:
# Register the cleaned Payments table in our SQL connection.
# This makes the current payments DataFrame available to SQL.
payments_clean = payments.copy()

payments_clean.to_sql(
    "payments",
    conn,
    index=False,
    if_exists="replace"
)

# Analyze repayment performance by loan type.
query = """
SELECT
    l.Loan_Type,

    -- Number of payment records
    COUNT(p.Payment_ID) AS Total_Payments,

    -- Number of payments that were late
    SUM(
        CASE
            WHEN p.Days_Late > 0 THEN 1
            ELSE 0
        END
    ) AS Late_Payments,

    -- Percentage of payments that were late
    ROUND(
        100.0 * SUM(
            CASE
                WHEN p.Days_Late > 0 THEN 1
                ELSE 0
            END
        ) / COUNT(p.Payment_ID),
        2
    ) AS Late_Payment_Rate,

    -- Average delay among late payments
    ROUND(
        AVG(
            CASE
                WHEN p.Days_Late > 0 THEN p.Days_Late
            END
        ),
        2
    ) AS Avg_Days_Late

FROM payments p

JOIN loans l
    ON p.Loan_ID = l.Loan_ID

GROUP BY l.Loan_Type

ORDER BY Late_Payment_Rate DESC;
"""

payment_analysis = pd.read_sql(query, conn)

payment_analysis

,Loan_Type,Total_Payments,Late_Payments,Late_Payment_Rate,Avg_Days_Late
0,Auto Loan,36817,27589,74.94,5.97
1,Personal Loan,63010,47126,74.79,5.87
2,Business Loan,34142,25480,74.63,5.98
3,Home Loan,40568,30256,74.58,5.87
4,Education Loan,19200,14272,74.33,5.97


In [ ]:
# Calculate loan processing time directly from the cleaned DataFrames.
# This avoids depending on the SQL copy of the tables.

processing = loan_applications_clean[
    ["Application_ID", "Loan_Type", "Application_Date", "Decision_Date"]
].copy()

# Calculate the number of days taken to make the application decision.
processing["Decision_Days"] = (
    processing["Decision_Date"] - processing["Application_Date"]
).dt.total_seconds() / (24 * 60 * 60)

# Keep only applications that have a corresponding loan.
processing = processing.merge(
    loans_clean[
        ["Application_ID", "Approval_Date", "Disbursement_Date"]
    ],
    on="Application_ID",
    how="inner"
)

# Calculate the number of days between approval and disbursement.
processing["Disbursement_Days"] = (
    processing["Disbursement_Date"] - processing["Approval_Date"]
).dt.total_seconds() / (24 * 60 * 60)

# Standardize Loan_Type before grouping.
# This ensures values such as "personal loan" and "Personal Loan"

processing["Loan_Type"] = (
    processing["Loan_Type"]
    .str.strip()
    .str.title()
)

# Calculate the final business metrics.
processing_analysis = (
    processing
    .groupby("Loan_Type")
    .agg(
        Avg_Decision_Days=("Decision_Days", "mean"),
        Avg_Disbursement_Days=("Disbursement_Days", "mean")
    )
    .round(2)
    .sort_values("Avg_Decision_Days", ascending=False)
)

processing_analysis

,Avg_Decision_Days,Avg_Disbursement_Days
Loan_Type,,
Personal Loan,7.62,15.41
Business Loan,7.61,15.39
Education Loan,7.57,15.76
Home Loan,7.46,15.38
Auto Loan,7.45,15.34


In [81]:
# Standardize Loan_Type values so that variations such as
# "personal loan" and "HOME LOAN" become the same category.

loan_applications_clean["Loan_Type"] = (
    loan_applications_clean["Loan_Type"]
    .str.strip()
    .str.title()
)

loans_clean["Loan_Type"] = (
    loans_clean["Loan_Type"]
    .str.strip()
    .str.title()
)

In [82]:
# Replace the existing SQL tables with the cleaned versions.
# This ensures our SQL analysis uses the corrected data.

loan_applications_clean.to_sql(
    "loan_applications",
    conn,
    index=False,
    if_exists="replace"
)

loans_clean.to_sql(
    "loans",
    conn,
    index=False,
    if_exists="replace"
)

14480

In [ ]:
# Business Question :
# What is the application approval rate by loan type?

query = """
SELECT
    Loan_Type,

    -- Total number of applications received
    COUNT(Application_ID) AS Total_Applications,

    -- Number of approved applications
    SUM(
        CASE
            WHEN Application_Status = 'Approved' THEN 1
            ELSE 0
        END
    ) AS Approved_Applications,

    -- Number of rejected applications
    SUM(
        CASE
            WHEN Application_Status = 'Rejected' THEN 1
            ELSE 0
        END
    ) AS Rejected_Applications,

    -- Percentage of applications that were approved
    ROUND(
        100.0 * SUM(
            CASE
                WHEN Application_Status = 'Approved' THEN 1
                ELSE 0
            END
        ) / COUNT(Application_ID),
        2
    ) AS Approval_Rate

FROM loan_applications

GROUP BY Loan_Type

ORDER BY Approval_Rate DESC;
"""

application_analysis = pd.read_sql(query, conn)

application_analysis

,Loan_Type,Total_Applications,Approved_Applications,Rejected_Applications,Approval_Rate
0,Home Loan,3942,3195,747,81.05
1,Auto Loan,3658,2952,706,80.70
2,Personal Loan,6790,5399,1391,79.51
3,Education Loan,1965,1518,447,77.25
4,Business Loan,3636,2757,879,75.83


Exporting the cleaned data

In [84]:
# ---------------------------------------------------------
# Export cleaned banking tables
# ---------------------------------------------------------
# Power BI will use these cleaned CSV files as its source.
# Keeping the cleaned data separate from data_raw is important:
# raw data remains untouched, while Power BI receives
# analyst-prepared data.
# ---------------------------------------------------------

customers_clean.to_csv(
    "../data_clean/customers_clean.csv",
    index=False
)

branches_clean.to_csv(
    "../data_clean/branches_clean.csv",
    index=False
)

loan_applications_clean.to_csv(
    "../data_clean/loan_applications_clean.csv",
    index=False
)

loans_clean.to_csv(
    "../data_clean/loans_clean.csv",
    index=False
)

payments_clean.to_csv(
    "../data_clean/payments_clean.csv",
    index=False
)

print("Cleaned banking tables exported successfully.")

Cleaned banking tables exported successfully.
